src/main/java/org/broadinstitute/hellbender/transformers/BQSRReadTransformer.java

```diff
@@ -101,75 +118,89 @@ public GATKRead apply(final GATKRead read) {
+            Arrays.fill(empiricalQualCovsArgs, null);  //clear the array
+            final int[] keySet = fullReadKeySet[offset];
 
-                        // recalibrated quality is bound between 1 and MAX_QUAL
-                        final byte recalibratedQual = QualityUtils.boundQual(MathUtils.fastRound(recalibratedQualDouble), RecalDatum.MAX_RECALIBRATED_Q_SCORE);
 
-                        // return the quantized version of the recalibrated quality
-                        final byte recalibratedQualityScore = quantizationInfo.getQuantizedQuals().get(recalibratedQual);
+            final RecalDatum empiricalQualQS = qualityScoreTable.get3(keySet[0], keySet[1], BASE_SUBSTITUTION_INDEX);
 
-                        quals[offset] = recalibratedQualityScore;
-                    }
+            for (int i = specialCovariateCount; i < totalCovariateCount; i++) {
+                if (keySet[i] >= 0) {
+                    empiricalQualCovsArgs[i - specialCovariateCount] = recalibrationTables.getTable(i).get4(keySet[0], keySet[1], keySet[i], BASE_SUBSTITUTION_INDEX);
                 }
             }
+            final double recalibratedQualDouble = hierarchicalBayesianQualityEstimate(epsilon, empiricalQualRG, empiricalQualQS, empiricalQualCovsArgs);
 
-            // finally update the base qualities in the read
-            ReadUtils.setBaseQualities(read, quals, errorModel);
+            quals[offset] = quantizedQuals.get(getRecalibratedQual(recalibratedQualDouble));
         }
+        read.setBaseQualities(quals);
         return read;
     }
 
-    public static double hierarchicalBayesianQualityEstimate( final double epsilon, final RecalDatum empiricalQualRG, final RecalDatum empiricalQualQS, final List<RecalDatum> empiricalQualCovs ) {
+    // recalibrated quality is bound between 1 and MAX_QUAL
+    private byte getRecalibratedQual(double recalibratedQualDouble) {
+        return boundQual(fastRound(recalibratedQualDouble), MAX_RECALIBRATED_Q_SCORE);
+    }
+
+    public static double hierarchicalBayesianQualityEstimate( final double epsilon,
+                                                              final RecalDatum empiricalQualRG,
+                                                              final RecalDatum empiricalQualQS,
+                                                              final RecalDatum... empiricalQualCovs ) {
         final double globalDeltaQ = ( empiricalQualRG == null ? 0.0 : empiricalQualRG.getEmpiricalQuality(epsilon) - epsilon );
-        final double deltaQReported = ( empiricalQualQS == null ? 0.0 : empiricalQualQS.getEmpiricalQuality(globalDeltaQ + epsilon) - (globalDeltaQ + epsilon) );
+
+        final double deltaQReported;
+        if (empiricalQualQS == null) {
+            deltaQReported = 0.0;
+        } else {
+            final double conditionalPrior1 = globalDeltaQ + epsilon;
+            deltaQReported = empiricalQualQS.getEmpiricalQuality(conditionalPrior1) - conditionalPrior1;
+        }
```

Comment:
Honestly, I like the original better.  At the expense of a single addition operation, this line got blown up into 7 lines.